src/ai/backend/logging/logger.py

```diff
@@ -0,0 +1,349 @@
+    file_handler = logging.handlers.RotatingFileHandler(
+        filename=drv_config["path"] / drv_config["filename"],
+        backupCount=drv_config["backup-count"],
+        maxBytes=drv_config["rotation-size"],
+        encoding="utf-8",
+    )
+    file_handler.setLevel(config["level"])
+    file_handler.setFormatter(CustomJsonFormatter(fmt))
+    return file_handler
+
+
+def log_worker(
+    logging_config: Mapping[str, Any],
+    parent_pid: int,
+    log_endpoint: str,
+    ready_event: threading.Event,
+    msgpack_options: MsgpackOptions,
+) -> None:
+    console_handler = None
+    file_handler = None
+    logstash_handler = None
+    graylog_handler = None
+
+    # For future references: when implementing new kind of logging adapters,
+    # make sure to adapt our custom `Formatter.formatException()` approach;
+    # Otherwise it won't print out EXCEPTION level log (along with the traceback).
+    if "console" in logging_config["drivers"]:
+        console_handler = setup_console_log_handler(logging_config)
+
+    if "file" in logging_config["drivers"]:
+        file_handler = setup_file_log_handler(logging_config)
+
+    if "logstash" in logging_config["drivers"]:
+        from .handler.logstash import LogstashHandler
+
+        drv_config = logging_config["logstash"]
+        logstash_handler = LogstashHandler(
+            endpoint=drv_config["endpoint"],
+            protocol=drv_config["protocol"],
+            ssl_enabled=drv_config["ssl-enabled"],
+            ssl_verify=drv_config["ssl-verify"],
+            myhost="hostname",  # TODO: implement
+        )
+        logstash_handler.setLevel(logging_config["level"])
+        logstash_handler.setFormatter(SerializedExceptionFormatter())
+
+    if "graylog" in logging_config["drivers"]:
+        from .handler.graylog import setup_graylog_handler
+
+        graylog_handler = setup_graylog_handler(logging_config)
+        assert graylog_handler is not None
+        graylog_handler.setFormatter(SerializedExceptionFormatter())
+
+    zctx = zmq.Context()
+    agg_sock = zctx.socket(zmq.PULL)
+    agg_sock.bind(log_endpoint)
+    ep_url = yarl.URL(log_endpoint)
+    if ep_url.scheme.lower() == "ipc":
+        os.chmod(ep_url.path, 0o777)
+    try:
+        ready_event.set()
+        while True:
+            data = agg_sock.recv()
+            if not data:
+                return
+            unpacked_data = msgpack.unpackb(data, **msgpack_options["unpack_opts"])
+            if not unpacked_data:
+                break
+            rec = logging.makeLogRecord(unpacked_data)
+            if rec is None:
+                break
+            if console_handler:
+                console_handler.emit(rec)
+            try:
+                if file_handler:
+                    file_handler.emit(rec)
+                if logstash_handler:
+                    logstash_handler.emit(rec)
+                    print("logstash")
```

Comment:
It seems maybe that the print statement may have been accidentally included as unnecessary code.